### ISSUE TITLE
Kotlin codegen: sealed classes for unions

### ISSUE THREAD

**Issue by benjamin-bader** [Microsoft/thrifty#184]
Idea:

```thrift
union Union {
  1: i32 Foo;
  2: i64 Bar;
  3: string Baz
}
```

```kotlin
sealed class Union {
  data class Foo(val value: Int) : Union()
  data class Bar(val value: Long) : Union()
  data class Baz(val value: string) : Union()

  class Builder {
    var foo: Int?
    var bar: Long?
    var baz: String?

    fun Foo(value: Int) = apply {
      this.foo = value
      this.bar = null
      this.baz = null
    }

    // etc

    fun build(): Union = when {
      foo != null -> Foo(foo)
      bar != null -> Bar(bar)
      baz != null -> Baz(baz)
      else -> throw AssertionError("unpossible")
    }
  }
}
```

**Comment by benjamin-bader** [comment]
Just a thought - it may make more sense to not use data classes for the leaves, and to have our own implementations for equals/hashcode/toString.  It would potentially reduce the method count imposed by this design.

**Comment by benjamin-bader** [comment]
As it turns out, I _didn't_ think of this - hzsweers proposed this a few months ago.  Credit where credit is due!

**Comment by benjamin-bader** [comment]
A significant edge case here turns out to be _empty unions_.  This is a union with no data members.  As it happens, Apache Thrift supports this nonsensical construct, but Kotlin sealed classes can't represent it.  We might fall back on a regular old data class instead, or even an object, in the case where a union has no fields.

**Comment by benjamin-bader** [comment]
Implemented in #253, including support for empty unions.
